File: backend/notification-service/app/services/notification_service.py
"""Notification Service"""
import logging
from typing import Optional
from datetime import datetime

from app.models.notification import Notification
from app.repositories.notification_repository import NotificationRepository

logger = logging.getLogger(__name__)


class NotificationService:
    """Service for notification business logic"""

    def __init__(self, repository: NotificationRepository):
        self.repository = repository

    async def create_notification(
        self,
        user_id: int,
        notification_type: str,
        channel: str,
        recipient: str,
        subject: str,
        content: str,
        status: str,
        metadata: Optional[str] = None,
        error_message: Optional[str] = None,
    ) -> Notification:
        """Create a new notification"""
        notification = Notification(
            user_id=user_id,
            notification_type=notification_type,
            channel=channel,
            recipient=recipient,
            subject=subject,
            content=content,
            status=status,
            metadata=metadata,
            error_message=error_message,
        )

        logger.info(
            f"Creating notification for user {user_id}: "
            f"type={notification_type}, channel={channel}, status={status}"
        )

        return await self.repository.create(notification)

    async def get_notification_by_id(self, notification_id: int) -> Optional[Notification]:
        """Get notification by ID"""
        logger.debug(f"Fetching notification {notification_id}")
        return await self.repository.get_by_id(notification_id)

    async def get_user_notifications(
        self, user_id: int, page: int = 1, page_size: int = 20
    ) -> tuple[list[Notification], int, int]:
        """Get paginated notifications for a user"""
        logger.debug(f"Fetching notifications for user {user_id}")
        skip = (page - 1) * page_size
        notifications, total = await self.repository.get_by_user_id(
            user_id, skip=skip, limit=page_size
        )
        total_pages = (total + page_size - 1) // page_size
        return notifications, total, total_pages

    async def get_user_notifications_by_type(
        self, user_id: int, notification_type: str
    ) -> list[Notification]:
        """Get notifications for a user by type"""
        logger.debug(f"Fetching {notification_type} notifications for user {user_id}")
        return await self.repository.get_by_user_and_type(user_id, notification_type)

    async def get_user_notifications_by_status(
        self, user_id: int, status: str, page: int = 1, page_size: int = 20
    ) -> tuple[list[Notification], int, int]:
        """Get paginated notifications for a user by status"""
        logger.debug(f"Fetching {status} notifications for user {user_id}")
        skip = (page - 1) * page_size
        notifications, total = await self.repository.get_by_user_and_status(
            user_id, status, skip=skip, limit=page_size
        )
        total_pages = (total + page_size - 1) // page_size
        return notifications, total, total_pages

    async def get_notifications_by_status(self, status: str) -> list[Notification]:
        """Get all notifications with a specific status"""
        logger.debug(f"Fetching all {status} notifications")
        return await self.repository.get_by_status(status)

    async def get_notifications_by_date_range(
        self, start: datetime, end: datetime
    ) -> list[Notification]:
        """Get notifications within a date range"""
        logger.debug(f"Fetching notifications between {start} and {end}")
        return await self.repository.get_by_date_range(start, end)

    async def count_user_notifications_by_status(self, user_id: int, status: str) -> int:
        """Count notifications for a user with a specific status"""
        logger.debug(f"Counting {status} notifications for user {user_id}")
        return await self.repository.count_by_user_and_status(user_id, status)

    async def mark_as_success(self, notification: Notification) -> Notification:
        """Mark notification as successful"""
        logger.info(f"Marking notification {notification.id} as SUCCESS")
        notification.status = "SUCCESS"
        return await self.repository.update(notification)

    async def mark_as_failed(
        self, notification: Notification, error_message: str
    ) -> Notification:
        """Mark notification as failed"""
        logger.warning(
            f"Marking notification {notification.id} as FAILED: {error_message}"
        )
        notification.status = "FAILED"
        notification.error_message = error_message
        return await self.repository.update(notification)

File: backend/notification-service/src/main/java/com/bookmyshow/notification/service/NotificationService.java
package com.bookmyshow.notification.service;

import com.bookmyshow.notification.entity.Notification;
import com.bookmyshow.notification.repository.NotificationRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Service
@Slf4j
public class NotificationService {

    private final NotificationRepository notificationRepository;

    @Autowired
    public NotificationService(NotificationRepository notificationRepository) {
        this.notificationRepository = notificationRepository;
    }

    @Transactional
    public Notification saveNotification(Notification notification) {
        log.debug("Saving notification for user: {}", notification.getUserId());
        return notificationRepository.save(notification);
    }

    @Transactional
    public Notification createNotification(Long userId, String notificationType, String channel,
                                          String recipient, String subject, String content, String status) {
        Notification notification = Notification.builder()
                .userId(userId)
                .notificationType(notificationType)
                .channel(channel)
                .recipient(recipient)
                .subject(subject)
                .content(content)
                .status(status)
                .build();

        log.info("Creating notification for user {}: type={}, channel={}, status={}",
                userId, notificationType, channel, status);
        return notificationRepository.save(notification);
    }

    @Transactional
    public Notification createNotificationWithMetadata(Long userId, String notificationType, String channel,
                                                      String recipient, String subject, String content,
                                                      String status, String metadata) {
        Notification notification = Notification.builder()
                .userId(userId)
                .notificationType(notificationType)
                .channel(channel)
                .recipient(recipient)
                .subject(subject)
                .content(content)
                .status(status)
                .metadata(metadata)
                .build();

        log.info("Creating notification with metadata for user {}: type={}, channel={}, status={}",
                userId, notificationType, channel, status);
        return notificationRepository.save(notification);
    }

    @Transactional(readOnly = true)
    public Page<Notification> getUserNotifications(Long userId, Pageable pageable) {
        log.debug("Fetching notifications for user: {}", userId);
        return notificationRepository.findByUserId(userId, pageable);
    }

    @Transactional(readOnly = true)
    public List<Notification> getUserNotificationsByType(Long userId, String notificationType) {
        log.debug("Fetching notifications for user {} with type: {}", userId, notificationType);
        return notificationRepository.findByUserIdAndNotificationType(userId, notificationType);
    }

    @Transactional(readOnly = true)
    public Page<Notification> getUserNotificationsByStatus(Long userId, String status, Pageable pageable) {
        log.debug("Fetching notifications for user {} with status: {}", userId, status);
        return notificationRepository.findByUserIdAndStatus(userId, status, pageable);
    }

    @Transactional(readOnly = true)
    public List<Notification> getNotificationsByStatus(String status) {
        log.debug("Fetching all notifications with status: {}", status);
        return notificationRepository.findByStatus(status);
    }

    @Transactional(readOnly = true)
    public List<Notification> getNotificationsByDateRange(LocalDateTime start, LocalDateTime end) {
        log.debug("Fetching notifications between {} and {}", start, end);
        return notificationRepository.findByCreatedAtBetween(start, end);
    }

    @Transactional(readOnly = true)
    public long countUserNotificationsByStatus(Long userId, String status) {
        log.debug("Counting notifications for user {} with status: {}", userId, status);
        return notificationRepository.countByUserIdAndStatus(userId, status);
    }

    @Transactional
    public Notification markAsSuccess(Notification notification) {
        notification.setStatus("SUCCESS");
        log.info("Marking notification {} as SUCCESS", notification.getId());
        return notificationRepository.save(notification);
    }

    @Transactional
    public Notification markAsFailed(Notification notification, String errorMessage) {
        notification.setStatus("FAILED");
        notification.setErrorMessage(errorMessage);
        log.warn("Marking notification {} as FAILED: {}", notification.getId(), errorMessage);
        return notificationRepository.save(notification);
    }
}
